### PR TITLE
Feature/who does it serve2

### LINKED ISSUE #108: Push complete 'Who-is-Served' flow
Reporter: AB

https://goo.gl/T5fKfs

### PR DESCRIPTION
**Pull request by jbeard4**
Better implementation for Who does it Serve scren. Fixes #108 

Should be merged after PR #34

### CODE REVIEW

**Comment by jbeard4** [comment]
Custom fields demo:

**Comment by Rovack** [comment]
Just looking at the gif, wasn't "who is served?" supposed to include more than just the age? I thought ages were one half of it, and the other was stuff like "vegetarian ", "kosher", etc, specific to the type of service.
I'd just make sure with Adam/Marco that that part doesn't still need to be reflected somehow (maybe still just a textbox).

**Comment by jbeard4** [comment]
I am pretty sure. See the design reference attached to #108
@noidsign @adambard1

**Comment by noidsign** [comment]
@jbeard4  "Who is served" was put on pause during MVP. As @Rovack mentioned, post MVP "Who is served" had to be contextual to the type of service selected, meaning that _food_ services probably would have different parameters from _shelter_ services, for example.

For reference: [https://www.figma.com/file/fdzJjvkSugVGnUa3zNdvMNsU/StreetTeam_source_of_truth?node-id=396%3A74]
[https://www.figma.com/file/fdzJjvkSugVGnUa3zNdvMNsU/StreetTeam_source_of_truth?node-id=547%3A4548]

Besides, another variable still to be consider was if gender had to be a separate question or if it would be part of this flow. All in all, this flow still needs quite a lot of work, and it comes a bit as a surprise that it's a P1, since open text field would do the work for now.

**Comment by jbeard4** [comment]
Hi @noidsign 

Thanks for following up. I am not sure if that first figma link worked, as it just took me to the style guide:

screen shot 2018-08-20 at 7 43 11 am

I understood from @adambard1 that this was a high priority. Let's plan to discuss so we can understand how to move forward.

**Comment by noidsign** [comment]
Yeah, the link is not correct. I don't understand why it shared that page. I'm gonna check on Adam's availability and send an invite for a quick session on this. Thank you Jake

**Comment by adambard1** [comment]
When is this getting pushed to prod pls?

**Comment by jbeard4** [comment]
I will wrap this up this week.

**Comment by jbeard4** [comment]
@Rovack Can you please review and merge this branch? Thanks.

**Comment by Rovack** [comment]
@jbeard4 No major comments, but looking at @noidsign's comments, it looks like last this was discussed, it wasn't at all clear we want to merge this from a product perspective. Has there been any follow-up to that?

File: src/app/serviceForm/whoDoesItServe/WhoDoesItServeEdit.jsx [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+import Header from '../../../components/header';
+import Button from '../../../components/button';
+import Selector from '../../../components/selector';
+import Input from '../../../components/input';
+import { SERVICE_GROUPS } from '../../../constants';
+import { formatLabel, isEditing } from './util'
+import './WhoDoesItServeEditForm.css';
+
+class ServiceOpeningHours extends Component {
```
[the remark this answers is not in the document]
Comment (jbeard4) [reply src/app/serviceForm/whoDoesItServe/WhoDoesItServeEdit.jsx]: Yes, class name should be updated. I will do this.

File: src/app/serviceForm/whoDoesItServe/WhoDoesItServeEdit.jsx [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    this.addCustomGroup = this.addCustomGroup.bind(this);
+    this.getCustomGroups = this.getCustomGroups.bind(this);
+
+    // serviceGroups -> { allAges: true, customMinAge, customMaxAge, name}
+    let serviceGroups = isEditing(this.props.value) ? [] : this.props.value;
+    this.state = {serviceGroups};
+  }
+
+  onServiceGroupClick(groupName, defaultMinAge, defaultMaxAge){
+    // toggle him
```
[the remark this answers is not in the document]
Comment (jbeard4) [reply src/app/serviceForm/whoDoesItServe/WhoDoesItServeEdit.jsx]: Fixed

**Comment by Rovack** [comment]
I don't know that there's currently any structure for this in the database.

But anyway, I don't have any comments from a code perspective, so it's approved from my side.
If there's any open product questions, I'm sure you guys will sort it out.

**Comment by jbeard4** [comment]
@noidsign @Rovack 

Here's an example of the JSON structure which is getting saved to the database, at the `services.who_does_it_serve` field:

```json
{
  "whoDoesItServe": [
    {
      "allAges": true,
      "name": "Everyone"
    },
    {
      "allAges": true,
      "minAge": 0,
      "maxAge": 11,
      "name": "Children"
    },
    {
      "allAges": false,
      "minAge": 12,
      "maxAge": 18,
      "name": "Teens"
    },
    {
      "allAges": false,
      "minAge": "12",
      "maxAge": "23",
      "name": "Veterans"
    },
    {
      "allAges": true,
      "minAge": 65,
      "name": "Seniors"
    },
    {
      "allAges": false,
      "minAge": "19",
      "maxAge": "120",
      "name": "Youths"
    },
    {
      "allAges": false,
      "name": "hello-world",
      "minAge": "0",
      "maxAge": "120"
    }
  ]
}
```

This field has type `json`, which is queryable, so we should be able to use this for eligibility later.

**Comment by noidsign** [comment]
Hey @jbeard4 this is great, thanks!
There is one thing I don't understand: in the examples above there is a boolean `allAges` that is set to `true` even when the service is not for any age (e.g. children 0-11). How are the `minAge` `maxAge` and `allAges` related?

Also, I wanted to ask you a bit more about how information can be retrieved? Also, at the moment it seems like a same `name` (e.g. Seniors) might have different parameters: is there a way to flag cases like this? Do we have a taxonomy doc easily accessible also by non developers? Thanks!

**Pull request merged**